### PR TITLE
Update task list component to make use of new experimental list

### LINKED ISSUE #6773: Refactor `<TaskList>`  and `getAllTasks` to support the new `<List>` component
This is dependent on the completion of #6772 

`<TaskList>` (located here: https://github.com/woocommerce/woocommerce-admin/blob/39b5cb0130c9ad0da2401f9f35110ae1b1d0c213/client/task-list/list.js#L29) is sort of a master/detail component. It controls a list of items and the displaying of a corresponding set up task. It is used in both setup task list and extended task list but won't be able to do the work of both in its current state.

It relies on `getAllTasks` located here: https://github.com/woocommerce/woocommerce-admin/blob/39b5cb0130c9ad0da2401f9f35110ae1b1d0c213/client/task-list/tasks.js#L46 

`getAllTasks` is technical debt, it won't scale well for the upcoming complexity we're adding to the task list.
This is negotiable but I'm proposing as part of this we:

1. Do away with `getAllTasks`. Create a base `<TaskItem>` component that covers the current use cases and also can support proposed new functionality discussed here: https://github.com/woocommerce/woocommerce-admin/issues/6639#issuecomment-816844512 

2. Create a component for each task we currently have. Use the `<TaskItem>` component internally. Get the state needed to determine if/what the task should do in each individual component. This will remove the need for `getAllTasks`. Because we use a `container` prop for the detail currently, there's a question of how we'll handle the master/detail part here. I don't have an answer for this yet, I'm hoping a solution will present itself as we get into it.

3. Also note that the new designs for the setup task list mentioned in `1` deviate from the functionality we'll provide for the extended task list, so this is a good time to also render a different kind of `<TaskItem>` component for the extended task list items, at this stage they are simplified and cut down to just providing hierarchy information as described in https://github.com/woocommerce/woocommerce-admin/issues/6639

4. We'll need to retain the existing containers for each task (the detail part) otherwise the refactor will be too cumbersome.

**Important note:** Instead of deleting `<TaskList>` straight away we could keep it for use in the extended task list so we can implement https://github.com/woocommerce/woocommerce-admin/issues/6638 without waiting on this refactoring work.

### PR DESCRIPTION
**Pull request by louwie17**
Fixes #6773 

Updated the task-list code to make use of the ExperimentalList that got merged in https://github.com/woocommerce/woocommerce-admin/pull/6787. I also converted the components to functional components.

### Screenshots

Screen Shot 2021-04-26 at 1 37 58 PM

### Detailed test instructions:

- Start a new WooCommerce store and finish the onboarding flow
- Smoke test the task list, the task list should be have the same way it currently does
     - Hovering over tasks, clicking on tasks, and opening tasks
     - It should trigger the correct tracks (tasklist_view, task_view, tasklist_click)
- Also smoke test the 3PD task list, by following the JS portion of [this tutorial](https://developer.woocommerce.com/2021/02/26/use-setup-tasks-to-provide-a-first-class-on-boarding-experience-for-merchants/), you can create a new extension using `npm run example -- --ext=add-task` 
- Set `isDismissable` to true for one of you tasks and click this, the animation should be correct as well.

### CODE REVIEW

**Comment by samueljseay** [comment]
@louwie17 I haven't tested this yet but the code looks pretty good to me! One thing I'll have to explore is whether there are different needs for a TaskItem component that we might expose externally, but I think we could look to extending or modifying this component as needed when the time comes to do that.

File: client/task-list/task-item.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
+/**
```
Comment (louwie17) [inline client/task-list/task-item.tsx:1]: This will likely get iterated up-on to accept the new task list functionality.

**Review by jeffstieler — APPROVED** [review]
Code looks good (some minor comments/questions), but I'm pre-approving. Nice work.

There is a visual regression with the dismiss button that needs to get taken care of before merging:

Screen Shot 2021-04-26 at 3 58 48 PM

**Review by psealock — COMMENTED** [review]
In https://github.com/woocommerce/woocommerce-admin/pull/6787 we spoke about moving this new component to `package/experimental`.

```js
import { ExperimentalListItem } from '@woocommerce/components';

// becomes

import { ListItem } from '@woocommerce/experimental';
```
The reason was for easier bundling by extensions and shift to deprecating the old component with the same name. I do think it should still be exported as it is by `@woocommerce/components` and re-exported by experimental as `List` and `ListItem`. What do you think? If that sounds good, the code here would obviously work, but maybe it would be best to import from experimental? We can always tackle in a follow up to keep this PR separate if you'd like

**Comment by louwie17** [comment]
> The reason was for easier bundling by extensions and shift to deprecating the old component with the same name. I do think it should still be exported as it is by @woocommerce/components and re-exported by experimental as List and ListItem. What do you think? If that sounds good, the code here would obviously work, but maybe it would be best to import from experimental? We can always tackle in a follow up to keep this PR separate if you'd like

I think that makes sense, @jeffstieler and I had talked about it, but where thinking it would require the Experimental List code to live in the experimental package.
I believe the above approach is indeed better, just exporting it as `List` and `ListItem` from the experimental package. I can add that change to this PR.

File: client/task-list/task-item.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import sanitizeHTML from '../lib/sanitize-html';
+
+type ListItemProps = {
+	title: string;
+	completed: boolean;
+	onClick: () => void;
+	isDismissable?: boolean;
+	onDismiss?: () => void;
+	additionalInfo?: string;
+	time: string;
```
[the remark this answers is not in the document]
Comment (louwie17) [reply client/task-list/task-item.tsx:24]: Yeah it should, thanks!

File: client/task-list/task-item.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+							</div>
+						) }
+						{ time && ! completed && (
+							<div className="woocommerce-task__estimated-time">
+								{ time }
+							</div>
+						) }
+					</Text>
+				</span>
+				<span className="woocommerce-task-list__item-content"></span>
```
[the remark this answers is not in the document]
Comment (louwie17) [reply client/task-list/task-item.tsx:82]: Thanks, it's not suppose to be there.

File: client/task-list/task-item.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+			</div>
+			{ onDismiss && isDismissable && ! completed && (
+				<div className="woocommerce-task-list__item-after">
+					<Button
+						data-testid={ `dismiss-button` }
+						isTertiary
+						onClick={ ( event: React.MouseEvent ) => {
```
[the remark this answers is not in the document]
Comment (louwie17) [reply client/task-list/task-item.tsx:89]: Given we are just using it for `preventDefault` it won't really matter, but I will add `| KeyboardEvent` to it, as it also contains `preventDefault`.

**Comment by louwie17** [comment]
@psealock I end up not exporting it from the experimental package, as it would cause a circular dependency (this actually caused nginx to break locally) The woocommerce/components package depends on the experimental package. I couldn't really think of a good way around that, aside from an absolute reference (which I didn't do, as I am not sure how Lerna handles that across packages).

I end up just importing it into our code as `{ __experimentalList as List } from '@woocommerce/components'`

> There is a visual regression with the dismiss button that needs to get taken care of before merging:

Thanks @jeffstieler this is fixed in https://github.com/woocommerce/woocommerce-admin/pull/6849/commits/4c25ccb62c0e6880a5a6ae0942347c039ba99352

**Comment by psealock** [comment]
>I end up not exporting it from the experimental package, as it would cause a circular dependency (this actually caused nginx to break locally) The woocommerce/components package depends on the experimental package

Oh right, I hadn't thought of that. What if the code moved over to `@woocommerce/experimental` ?

**Comment by louwie17** [comment]
> Oh right, I hadn't thought of that. What if the code moved over to `@woocommerce/experimental` ?

Yeah we could still do that, I can push up a separate PR for that and see how it fits.

**Pull request merged**